File: framework/php/src/Elevation.php
<?php
declare(strict_types=1);
namespace BetterPortal\Protocol;

/** Protocol helper only: the host MUST verify the JWT, scope and permissions first. */
final class Elevation
{
    /** Returns null when satisfied, otherwise a 401 challenge. Invalid requirements fail closed. */
    public static function challenge(?array $verifiedUser, array $requirement, ?int $now = null): ?array
    {
        $minimum = $requirement['minimum'] ?? null;
        $age = $requirement['maxAgeSeconds'] ?? null;
        if (!in_array($minimum, ['confirm', 'mfa'], true) || ($age !== null && (!is_int($age) || $age < 1))) throw new \InvalidArgumentException('Invalid elevation requirement');
        $now ??= time();
        if ($verifiedUser === null || ($verifiedUser['tokenType'] ?? null) !== 'access' || !isset($verifiedUser['tenantId'], $verifiedUser['appId'], $verifiedUser['exp']) || $verifiedUser['exp'] <= $now) throw new \RuntimeException('Verified human authentication required');
        $e = $verifiedUser['elevation'] ?? null;
        if (is_array($e) && in_array($e['assurance'] ?? null, ['confirmed', 'mfa'], true)
            && is_int($e['verifiedAt'] ?? null) && is_int($e['expiresAt'] ?? null)
            && 0 <= $e['verifiedAt'] && $e['verifiedAt'] <= $now && $now < $e['expiresAt'] && $e['expiresAt'] <= $verifiedUser['exp']
            && ($minimum !== 'mfa' || $e['assurance'] === 'mfa') && ($age === null || $now - $e['verifiedAt'] <= $age)) return null;
        $challenge = ['version' => 1, 'tenantId' => $verifiedUser['tenantId'], 'appId' => $verifiedUser['appId'], 'minimum' => $minimum];
        $authenticate = 'Bearer error="insufficient_user_authentication", acr_values="urn:betterportal:' . ($minimum === 'mfa' ? 'mfa' : 'confirmed') . '"';
        if ($age !== null) { $challenge['maxAgeSeconds'] = $age; $authenticate .= ', max_age="' . $age . '"'; }
        return ['status' => 401, 'headers' => ['WWW-Authenticate' => $authenticate, 'BP-Auth-Challenge' => json_encode($challenge, JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES), 'Cache-Control' => 'no-store'], 'body' => ['error' => 'insufficient_user_authentication', ...$challenge]];
    }
}

File: framework/go/elevation.go
package protocol

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"time"
)

// ElevationRequirement is additional assurance, never a substitute for permissions.
type ElevationRequirement struct {
	Minimum       string `json:"minimum"`
	MaxAgeSeconds *int64 `json:"maxAgeSeconds,omitempty"`
}
type ElevationClaims struct {
	Assurance  string `json:"assurance"`
	VerifiedAt int64  `json:"verifiedAt"`
	ExpiresAt  int64  `json:"expiresAt"`
}

// VerifiedAccess must come from the host's JWT verifier and app-scope check.
// This protocol package does not implement JWT verification or a route server.
type VerifiedAccess struct {
	TenantID  string
	AppID     string
	TokenType string
	ExpiresAt int64
	Elevation *ElevationClaims
}
type ElevationRequired struct{ Headers http.Header }

func (e *ElevationRequired) Error() string { return "insufficient_user_authentication" }
func (e *ElevationRequired) Status() int   { return http.StatusUnauthorized }

// RequireElevation fails closed. Call after authentication and permissions, before effects.
func RequireElevation(user *VerifiedAccess, requirement ElevationRequirement, now time.Time) error {
	if requirement.Minimum != "confirm" && requirement.Minimum != "mfa" || requirement.MaxAgeSeconds != nil && *requirement.MaxAgeSeconds < 1 {
		return errors.New("invalid elevation requirement")
	}
	if user == nil || user.TenantID == "" || user.AppID == "" || user.TokenType != "access" || user.ExpiresAt <= now.Unix() {
		return errors.New("verified human authentication required")
	}
	e := user.Elevation
	if e != nil && (e.Assurance == "confirmed" || e.Assurance == "mfa") && e.VerifiedAt >= 0 && e.VerifiedAt <= now.Unix() && now.Unix() < e.ExpiresAt && e.ExpiresAt <= user.ExpiresAt && (requirement.Minimum != "mfa" || e.Assurance == "mfa") && (requirement.MaxAgeSeconds == nil || now.Unix()-e.VerifiedAt <= *requirement.MaxAgeSeconds) {
		return nil
	}
	challenge := map[string]any{"version": 1, "tenantId": user.TenantID, "appId": user.AppID, "minimum": requirement.Minimum}
	acr := "confirmed"
	if requirement.Minimum == "mfa" {
		acr = "mfa"
	}
	authenticate := fmt.Sprintf(`Bearer error="insufficient_user_authentication", acr_values="urn:betterportal:%s"`, acr)
	if requirement.MaxAgeSeconds != nil {
		challenge["maxAgeSeconds"] = *requirement.MaxAgeSeconds
		authenticate += fmt.Sprintf(`, max_age="%d"`, *requirement.MaxAgeSeconds)
	}
	encoded, err := json.Marshal(challenge)
	if err != nil {
		return err
	}
	headers := http.Header{}
	headers.Set("WWW-Authenticate", authenticate)
	headers.Set("BP-Auth-Challenge", string(encoded))
	headers.Set("Cache-Control", "no-store")
	return &ElevationRequired{Headers: headers}
}
